eth_account/ledger.py

```diff
@@ -0,0 +1,124 @@
+import struct
+
+import ledgerblue.comm
+
+ETH_DERIVATION_PATH_PREFIX = "44'/60'/0'/"
+
+class LedgerAccount:
+    """
+    Ledger Ethereum App Protocol Spec is located at:
+    <https://github.com/LedgerHQ/blue-app-eth/blob/master/doc/ethapp.asc>
+
+    References:
+    - https://github.com/LedgerHQ/blue-app-eth/blob/master/getPublicKey.py
+    - https://github.com/bargst/pyethoff/blob/master/tx_sign.py
+
+    """
+
+    def __init__(self, offset=None, address='')
+        self.dongle = ledgerblue.comm.getDongle(debug=False)
+
+        if offset is None and :
+            # Convert an address to an offset
+            if address == '':
+                raise Exception('Invalid offset and address provided')
+            else:
+                #TODO check address validity
+                self.offset = self.get_offset(address)
+                self.address = address
+        else:
+            self.offset = offset
+            self.address = self.get_addresses(offset)
+
+
+    def _parse_bip32_path(self, offset):
+        """
+        Convert an offset to a bytes payload to be sent to the ledger 
+        representing bip32 derivation path.
+        """
+        path = ETH_DERIVATION_PATH_PREFIX + str(offset)
+        result = bytes()
+        elements = path.split('/')
+        for pathElement in elements:
+            element = pathElement.split("'")
+            if len(element) == 1:
+                result = result + struct.pack(">I", int(element[0]))            
+            else:
+                result = result + struct.pack(">I", 0x80000000 | int(element[0]))
+        return result
+
+    def get_addresses(self, offset):
+        """
+        Query the ledger device for a list of ethereum address.
+        Offset is the number in the HD wallet tree
+        """
+        donglePath = self._parse_bip32_path(offset)
+
+        apdu =  bytes.fromhex('e0020000') # https://github.com/LedgerHQ/blue-app-eth/blob/master/doc/ethapp.asc#get-eth-public-address
+        apdu += bytes([len(donglePath) + 1])
+        apdu += bytes([len(donglePath) // 4])
+        apdu += donglePath
+
+        result = self.dongle.exchange(apdu, timeout=60)
+
+        # Parse result
+        offset = 1 + result[0]
+        address = result[offset + 1 : offset + 1 + result[offset]]
+
+        return f'0x{address.decode()}'
+
+    def get_offset(self, address):
+        """
+        Convert an address to the HD wallet tree offset
+        """
+        offset = 0
+        while address != self.get_addresses(offset):
+            offset += 1
```

Comment:
Instead of this mutation pattern, how about:

```python
import itertools

for offset in itertools.count():
    ...
```

Also, lets set an upper bound on this to some reasonable value (like 20?).

Reply:
> Also, lets set an upper bound on this to some reasonable value (like 20?).

I'm not sure we can define what maximum number of addresses should be allowed against a wallet. I guess we can start with something like: how many addresses can we check in one second? At the least, we would need an escape hatch for people who might have tons of addresses, like:

```py
    def get_offset(self, address, maximum_search_length=100):
```

Reply:
Good plan, like the kwarg.  how about `search_limit`?  We should probably also allow for specifying the starting offset value.